java-checks/src/main/java/org/sonar/java/checks/ThreadAsRunnableArgumentCheck.java

```diff
@@ -19,85 +19,131 @@
 import org.sonar.plugins.java.api.semantic.Symbol;
 import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
 import org.sonar.plugins.java.api.tree.ExpressionTree;
-import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
 import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewArrayTree;
 import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
 import org.sonar.plugins.java.api.tree.Tree;
-import java.text.MessageFormat;
-import java.util.Arrays;
-import java.util.List;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.plugins.java.api.tree.YieldStatementTree;
 
 @Rule(key = "S2438")
 public class ThreadAsRunnableArgumentCheck extends IssuableSubscriptionVisitor {
 
+  private static final String RUNNABLE_TYPE = "java.lang.Runnable";
+  private static final String THREAD_TYPE = "java.lang.Thread";
+  private static final String RUNNABLE_ARRAY_TYPE = RUNNABLE_TYPE + "[]";
+  private static final String THREAD_ARRAY_TYPE = THREAD_TYPE + "[]";
+
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return Arrays.asList(Tree.Kind.NEW_CLASS, Tree.Kind.METHOD_INVOCATION);
+    return List.of(Tree.Kind.VARIABLE, Tree.Kind.RETURN_STATEMENT, Tree.Kind.YIELD_STATEMENT, Tree.Kind.ASSIGNMENT,
+      Tree.Kind.METHOD_INVOCATION, Tree.Kind.NEW_CLASS, Tree.Kind.NEW_ARRAY);
   }
 
   @Override
   public void visitNode(Tree tree) {
-    List<ExpressionTree> arguments;
-    Symbol.MethodSymbol methodSymbol;
-    if (tree.is(Tree.Kind.NEW_CLASS)) {
-      NewClassTree nct = (NewClassTree) tree;
-      methodSymbol = nct.methodSymbol();
-      arguments = nct.arguments();
-    } else {
-      MethodInvocationTree mit = (MethodInvocationTree) tree;
-      methodSymbol = mit.methodSymbol();
-      arguments = mit.arguments();
+    switch(tree.kind()) {
+      case VARIABLE -> visitVariable((VariableTree) tree);
+      case RETURN_STATEMENT -> visitReturnStatement((ReturnStatementTree) tree);
+      case YIELD_STATEMENT -> visitYieldStatement((YieldStatementTree) tree);
+      case NEW_ARRAY -> visitNewArray((NewArrayTree) tree);
+      case ASSIGNMENT -> {
+        var assignment = (AssignmentExpressionTree) tree;
+        checkTypeCoercion(assignment.variable().symbolType(), assignment.expression());
+      }
+      case METHOD_INVOCATION -> {
+        var invacation = (MethodInvocationTree) tree;
+        visitInvocation(invacation.methodSymbol(), invacation.arguments());
+      }
+      case NEW_CLASS -> {
+        var invacation = (NewClassTree) tree;
+        visitInvocation(invacation.methodSymbol(), invacation.arguments());
+      }
+      default -> throw new IllegalArgumentException();
```

Comment:
IMO, this is dead code that we won't be able to test and we shouldn't add this line. I know we have a rule (S131 I believe) that will complain, but I don't think it really makes sense in contexts like this and would consider it a FP. Not really convinced that it is useful enough as defensive programming here either. I'm going to leave the final decision up to you though.

Reply:
Removed it. Now violating Joshua Bloch 6:36. May the lord forgive me.